src/watched_tasks.rs

```diff
@@ -0,0 +1,282 @@
+use std::future::Future;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use std::thread;
+
+use anyhow::{anyhow, Context as AnyhowContext, Result};
+use async_std::channel::{bounded, Receiver};
+use async_std::task;
+use log::info;
+
+type TaskResult = Result<()>;
+type TaskHandle = task::JoinHandle<TaskResult>;
+
+struct ThreadHandle {
+    handle: Option<thread::JoinHandle<TaskResult>>,
+    exit_notification: Option<Receiver<()>>,
+}
+
+pub struct WatchedTasksBuilder {
+    tasks: Vec<TaskHandle>,
+    threads: Vec<ThreadHandle>,
+}
+
+pub struct WatchedTasks {
+    tasks: Vec<TaskHandle>,
+    threads: Vec<ThreadHandle>,
+}
+
+impl ThreadHandle {
+    fn new<F>(name: String, function: F) -> Result<Self>
+    where
+        F: FnOnce() -> TaskResult + Send + 'static,
+    {
+        let (tx, exit_notification) = bounded(1);
+
+        let handle = thread::Builder::new().name(name).spawn(move || {
+            let res = function();
+            tx.try_send(())
+                .expect("Failed to send thread completion notification");
+            res
+        })?;
+
+        Ok(Self {
+            handle: Some(handle),
+            exit_notification: Some(exit_notification),
+        })
+    }
+
+    fn name(&self) -> Option<&str> {
+        self.handle
+            .as_ref()
+            .and_then(|handle| handle.thread().name())
+    }
+}
+
+impl Future for ThreadHandle {
+    type Output = TaskResult;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        if let Some(exit_notification) = self.exit_notification.as_mut() {
+            match Pin::new(&mut exit_notification.recv()).poll(cx) {
+                Poll::Pending => {
+                    return Poll::Pending;
+                }
+                Poll::Ready(Err(e)) => {
+                    self.exit_notification = None;
+                    return Poll::Ready(Err(e.into()));
+                }
+                Poll::Ready(Ok(())) => {
+                    self.exit_notification = None;
+                }
+            }
+        } else {
+            return Poll::Ready(Err(anyhow!(
+                "ThreadHandle was already polled to completion"
+            )));
+        }
+
+        let handle = self.handle.take().expect("Failed to get thread handle");
```

Comment:
```suggestion
        let handle = self
            .handle
            .take()
            .with_context(|| "Failed to get thread handle")?;
```

Reply:
Wouldn't work here as the `Future:poll()` implementation has to return a `Poll<>`, but I came up with another solution:

https://github.com/linux-automation/tacd/blob/69d52f700eb62493607edb6786486a6796b2d13a/src/watched_tasks.rs#L99-L108

Reply:
Yes, I've seen this change. I'm not that familiar with async Rust (yet) but from my understanding it seems sound.